File: src/llm_cal/performance/compute.py
"""Performance modeling for prefill latency and decode throughput.

FORMULAS — with sources. See docs/methodology.md for the full audit.

Prefill (compute-bound):
    FLOPs = 2 × params × input_tokens
    latency = FLOPs / (peak_TFLOPS × num_gpus × utilization × 1e12)

    Source: Kaplan et al. 2020, "Scaling Laws for Neural Language Models".
    The "2" factor is the forward-pass cost per param per token, a standard
    approximation in transformer inference literature.

Decode (memory-bandwidth-bound):
    per_token_time = weight_bytes_per_gpu / (memory_bandwidth × utilization)
    tokens_per_second = memory_bandwidth × utilization / weight_bytes_per_gpu

    Source: Kwon et al. SOSP 2023 "Efficient Memory Management for Large
    Language Model Serving with PagedAttention"; NVIDIA "Mastering LLM
    Techniques: Inference Optimization" (2023 technical blog).

UTILIZATION FACTORS (all empirical, ALL user-overridable):
  - Prefill 40% — midpoint of vLLM-reported 30-50% MFU on H100
  - Decode BW 50% — midpoint of NVIDIA/vLLM-reported 40-65% achieved bandwidth
  - Cluster comm 90% — typical NCCL AllReduce efficiency at TP=8 on NVLink
  - Concurrency degradation 1.0 (no degradation by default)
    This is the most uncertain factor. Prior versions defaulted to 1.5
    (borrowed from an LLM-generated report), which was NOT from a primary
    source. v0.1 defaults to 1.0 (honest baseline) and exposes the knob
    so users can dial in whatever their engine actually achieves.

MoE "active" vs "total":
    Strictly, MoE decode only reads the active experts per token. The
    ratio used here is a rough approximation:
        active_ratio ≈ (experts_per_tok + shared_experts) / (routed + shared)
    This UNDERESTIMATES active weight because attention + embeddings are
    always active (not just experts). For a more accurate number, use the
    model card's stated "total / active" figure if available. The
    "active-only" throughput is labeled "optimistic" for this reason.
"""

from __future__ import annotations

from dataclasses import dataclass

from llm_cal.architecture.profile import ArchitectureProfile
from llm_cal.hardware.loader import GPUSpec
from llm_cal.output.labels import AnnotatedValue, Label

# Empirical defaults. All user-overridable via CLI.
DEFAULT_PREFILL_UTILIZATION = 0.40
DEFAULT_DECODE_BW_UTILIZATION = 0.50
DEFAULT_CLUSTER_COMM_EFFICIENCY = 0.90
# Honest baseline. Previously 1.5, borrowed from an LLM-generated report —
# that had no primary source, so we reset to 1.0. Users who observe actual
# degradation on their engine should dial this up via CLI.
DEFAULT_CONCURRENCY_DEGRADATION = 1.0


@dataclass(frozen=True)
class PrefillEstimate:
    total_flops: AnnotatedValue[int]  # [estimated] 2 * params * input_tokens
    peak_effective_tflops: AnnotatedValue[float]  # TFLOPS × utilization
    latency_ms: AnnotatedValue[float]
    utilization: float  # the factor used (for provenance)


@dataclass(frozen=True)
class DecodeEstimate:
    active_weight_bytes_per_gpu: AnnotatedValue[int]
    per_gpu_tokens_per_sec: AnnotatedValue[float]
    cluster_tokens_per_sec: AnnotatedValue[float]  # after comm efficiency
    bw_utilization: float
    cluster_comm_efficiency: float
    moe_active_weight_bytes_per_gpu: AnnotatedValue[int] | None = None
    moe_active_tokens_per_sec: AnnotatedValue[float] | None = None


def estimate_prefill(
    profile: ArchitectureProfile,
    total_params: int,
    gpu: GPUSpec,
    num_gpus: int,
    input_tokens: int,
    utilization: float = DEFAULT_PREFILL_UTILIZATION,
) -> PrefillEstimate:
    """Estimate single-request prefill latency.

    Based on compute: FLOPs = 2 × params × tokens; latency = FLOPs / effective_FLOPS.
    """
    flops = 2 * total_params * input_tokens
    # TP distributes compute, so aggregate TFLOPS = num_gpus × per-card × util
    aggregate_tflops = gpu.fp16_tflops * num_gpus * utilization
    # Guard against zero
    if aggregate_tflops <= 0 or total_params <= 0 or input_tokens <= 0:
        return PrefillEstimate(
            total_flops=AnnotatedValue(0, Label.UNKNOWN, source="insufficient inputs"),
            peak_effective_tflops=AnnotatedValue(0.0, Label.UNKNOWN),
            latency_ms=AnnotatedValue(0.0, Label.UNKNOWN),
            utilization=utilization,
        )
    latency_s = flops / (aggregate_tflops * 1e12)
    latency_ms = latency_s * 1000.0

    return PrefillEstimate(
        total_flops=AnnotatedValue(
            flops,
            Label.ESTIMATED,
            source=f"2 × {total_params:,} params × {input_tokens:,} tokens",
        ),
        peak_effective_tflops=AnnotatedValue(
            aggregate_tflops,
            Label.ESTIMATED,
            source=f"{gpu.fp16_tflops} × {num_gpus} GPUs × {utilization:.0%} util",
        ),
        latency_ms=AnnotatedValue(
            latency_ms,
            Label.ESTIMATED,
            source=(f"{flops:.2e} FLOPs / ({aggregate_tflops:.1f} effective TFLOPS × 1e12)"),
        ),
        utilization=utilization,
    )


def _nvlink_efficiency(gpu: GPUSpec, num_gpus: int) -> float:
    """Multiplier on cluster comm efficiency reflecting NVLink bandwidth.

    Single-GPU has no TP all-reduce, so no penalty. H100 / B200 / H200 / A100-
    SXM4 with full NVLink (>=900 GB/s aggregate, dropped to 600 for A100) get
    ~1.0. Restricted-NVLink variants (H800: 400 GB/s, half of H100) pay ~8%.
    PCIe-only cards (L40S, RTX) with no NVLink pay 20%.
    """
    if num_gpus <= 1:
        return 1.0
    nvlink = gpu.nvlink_bandwidth_gbps or 0
    if nvlink >= 900:
        return 1.0
    if nvlink <= 0:
        return 0.80
    return 0.85 + 0.15 * (nvlink / 900.0)


def estimate_decode(
    profile: ArchitectureProfile,
    total_weight_bytes: int,
    gpu: GPUSpec,
    num_gpus: int,
    bw_utilization: float = DEFAULT_DECODE_BW_UTILIZATION,
    cluster_comm_efficiency: float = DEFAULT_CLUSTER_COMM_EFFICIENCY,
    moe_active_params_ratio: float | None = None,
) -> DecodeEstimate:
    """Estimate decode tokens/second.

    Decode is memory-bandwidth-bound: per-token time = weight_bytes / bw.
    Under TP, weights split across ranks, so per-GPU weight bytes = total / N.

    If the model is MoE and moe_active_params_ratio is given (e.g. 0.3 for
    active/total), we ALSO report an optimistic "active only" throughput.
    """
    if gpu.memory_bandwidth_gbps is None or gpu.memory_bandwidth_gbps <= 0:
        _unknown = AnnotatedValue(
            0, Label.UNKNOWN, source="GPU memory_bandwidth_gbps not in database"
        )
        _unknown_f = AnnotatedValue(
            0.0, Label.UNKNOWN, source="GPU memory_bandwidth_gbps not in database"
        )
        return DecodeEstimate(
            active_weight_bytes_per_gpu=_unknown,
            per_gpu_tokens_per_sec=_unknown_f,
            cluster_tokens_per_sec=_unknown_f,
            bw_utilization=bw_utilization,
            cluster_comm_efficiency=cluster_comm_efficiency,
        )

    bw_bytes_per_s = gpu.memory_bandwidth_gbps * 1e9  # GB/s → bytes/s
    effective_bw = bw_bytes_per_s * bw_utilization
    weight_per_gpu = max(1, total_weight_bytes // num_gpus)
    per_gpu_tps = effective_bw / weight_per_gpu
    # Cluster-level: per-GPU × N × comm_efficiency × NVLink-aware penalty.
    # NVLink penalty captures TP all-reduce overhead on cards with restricted
    # interconnect (H800, PCIe-only). Single-GPU is unaffected.
    nvlink_eff = _nvlink_efficiency(gpu, num_gpus)
    effective_comm_eff = cluster_comm_efficiency * nvlink_eff
    cluster_tps = per_gpu_tps * num_gpus * effective_comm_eff

    # MoE active-only optimistic view
    moe_active_weight: AnnotatedValue[int] | None = None
    moe_active_tps: AnnotatedValue[float] | None = None
    if profile.is_moe and moe_active_params_ratio is not None and moe_active_params_ratio > 0:
        active_bytes = int(weight_per_gpu * moe_active_params_ratio)
        moe_active_weight = AnnotatedValue(
            active_bytes,
            Label.ESTIMATED,
            source=f"{weight_per_gpu:,} × {moe_active_params_ratio:.3f} (active/total ratio)",
        )
        if active_bytes > 0:
            active_per_gpu_tps = effective_bw / active_bytes
            active_cluster_tps = active_per_gpu_tps * num_gpus * effective_comm_eff
            moe_active_tps = AnnotatedValue(
                active_cluster_tps,
                Label.ESTIMATED,
                source=(
                    f"optimistic MoE active-only: effective_bw / {active_bytes:,} × "
                    f"{num_gpus} × {effective_comm_eff:.3f}"
                ),
            )

    return DecodeEstimate(
        active_weight_bytes_per_gpu=AnnotatedValue(
            weight_per_gpu,
            Label.ESTIMATED,
            source=f"{total_weight_bytes:,} bytes / {num_gpus} TP ranks",
        ),
        per_gpu_tokens_per_sec=AnnotatedValue(
            per_gpu_tps,
            Label.ESTIMATED,
            source=(
                f"{gpu.memory_bandwidth_gbps} GB/s × {bw_utilization:.0%} util / "
                f"{weight_per_gpu:,} weight bytes"
            ),
        ),
        cluster_tokens_per_sec=AnnotatedValue(
            cluster_tps,
            Label.ESTIMATED,
            source=(
                f"per-GPU × {num_gpus} GPUs × {cluster_comm_efficiency:.0%} comm × "
                f"{nvlink_eff:.3f} NVLink penalty (NVLink={gpu.nvlink_bandwidth_gbps or 0} GB/s)"
            ),
        ),
        bw_utilization=bw_utilization,
        cluster_comm_efficiency=cluster_comm_efficiency,
        moe_active_weight_bytes_per_gpu=moe_active_weight,
        moe_active_tokens_per_sec=moe_active_tps,
    )

File: crates/llm-infer-cal-core/src/performance/compute.rs
use crate::architecture::profile::ArchitectureProfile;
use crate::hardware::loader::GPUSpec;
use crate::output::labels::{AnnotatedValue, Label};

pub const DEFAULT_PREFILL_UTILIZATION: f64 = 0.40;
pub const DEFAULT_DECODE_BW_UTILIZATION: f64 = 0.50;
pub const DEFAULT_CLUSTER_COMM_EFFICIENCY: f64 = 0.90;
pub const DEFAULT_CONCURRENCY_DEGRADATION: f64 = 1.0;

#[derive(Clone, Debug, PartialEq)]
pub struct PrefillEstimate {
    pub total_flops: AnnotatedValue<u64>,
    pub peak_effective_tflops: AnnotatedValue<f64>,
    pub latency_ms: AnnotatedValue<f64>,
    pub utilization: f64,
}

#[derive(Clone, Debug, PartialEq)]
pub struct DecodeEstimate {
    pub active_weight_bytes_per_gpu: AnnotatedValue<u64>,
    pub per_gpu_tokens_per_sec: AnnotatedValue<f64>,
    pub cluster_tokens_per_sec: AnnotatedValue<f64>,
    pub bw_utilization: f64,
    pub cluster_comm_efficiency: f64,
    pub moe_active_weight_bytes_per_gpu: Option<AnnotatedValue<u64>>,
    pub moe_active_tokens_per_sec: Option<AnnotatedValue<f64>>,
}

pub fn estimate_prefill(
    _profile: &ArchitectureProfile,
    total_params: u64,
    gpu: &GPUSpec,
    num_gpus: u64,
    input_tokens: u64,
    utilization: f64,
) -> PrefillEstimate {
    let flops = 2 * total_params * input_tokens;
    let aggregate_tflops = gpu.fp16_tflops * num_gpus as f64 * utilization;

    if aggregate_tflops <= 0.0 || total_params == 0 || input_tokens == 0 {
        return PrefillEstimate {
            total_flops: AnnotatedValue::new(0, Label::Unknown, Some("insufficient inputs")),
            peak_effective_tflops: AnnotatedValue::new(0.0, Label::Unknown, None),
            latency_ms: AnnotatedValue::new(0.0, Label::Unknown, None),
            utilization,
        };
    }

    let latency_ms = flops as f64 / (aggregate_tflops * 1e12) * 1000.0;
    let total_source = format!(
        "2 × {} params × {} tokens",
        fmt_u64(total_params),
        fmt_u64(input_tokens)
    );
    let peak_source = format!(
        "{:.1} × {} GPUs × {:.0}% util",
        gpu.fp16_tflops,
        num_gpus,
        utilization * 100.0
    );
    let latency_source = format!(
        "{} FLOPs / ({aggregate_tflops:.1} effective TFLOPS × 1e12)",
        fmt_scientific(flops)
    );

    PrefillEstimate {
        total_flops: AnnotatedValue::new(flops, Label::Estimated, Some(&total_source)),
        peak_effective_tflops: AnnotatedValue::new(
            aggregate_tflops,
            Label::Estimated,
            Some(&peak_source),
        ),
        latency_ms: AnnotatedValue::new(latency_ms, Label::Estimated, Some(&latency_source)),
        utilization,
    }
}

pub fn nvlink_efficiency(gpu: &GPUSpec, num_gpus: u64) -> f64 {
    if num_gpus <= 1 {
        return 1.0;
    }
    let nvlink = gpu.nvlink_bandwidth_gbps;
    if nvlink >= 900 {
        return 1.0;
    }
    if nvlink == 0 {
        return 0.80;
    }
    0.85 + 0.15 * (nvlink as f64 / 900.0)
}

pub fn estimate_decode(
    profile: &ArchitectureProfile,
    total_weight_bytes: u64,
    gpu: &GPUSpec,
    num_gpus: u64,
    bw_utilization: f64,
    cluster_comm_efficiency: f64,
    moe_active_params_ratio: Option<f64>,
) -> DecodeEstimate {
    let Some(memory_bandwidth_gbps) = gpu.memory_bandwidth_gbps else {
        return unknown_decode(bw_utilization, cluster_comm_efficiency);
    };
    if memory_bandwidth_gbps == 0 {
        return unknown_decode(bw_utilization, cluster_comm_efficiency);
    }

    let num_gpus = num_gpus.max(1);
    let bw_bytes_per_s = memory_bandwidth_gbps as f64 * 1e9;
    let effective_bw = bw_bytes_per_s * bw_utilization;
    let weight_per_gpu = (total_weight_bytes / num_gpus).max(1);
    let per_gpu_tps = effective_bw / weight_per_gpu as f64;
    let nvlink_eff = nvlink_efficiency(gpu, num_gpus);
    let effective_comm_eff = cluster_comm_efficiency * nvlink_eff;
    let cluster_tps = per_gpu_tps * num_gpus as f64 * effective_comm_eff;

    let mut moe_active_weight = None;
    let mut moe_active_tps = None;
    if profile.is_moe() {
        if let Some(ratio) = moe_active_params_ratio.filter(|ratio| *ratio > 0.0) {
            let active_bytes = (weight_per_gpu as f64 * ratio) as u64;
            let weight_source = format!(
                "{} × {ratio:.3} (active/total ratio)",
                fmt_u64(weight_per_gpu)
            );
            moe_active_weight = Some(AnnotatedValue::new(
                active_bytes,
                Label::Estimated,
                Some(&weight_source),
            ));
            if active_bytes > 0 {
                let active_cluster_tps =
                    (effective_bw / active_bytes as f64) * num_gpus as f64 * effective_comm_eff;
                let tps_source = format!(
                    "optimistic MoE active-only: effective_bw / {} × {num_gpus} × {effective_comm_eff:.3}",
                    fmt_u64(active_bytes)
                );
                moe_active_tps = Some(AnnotatedValue::new(
                    active_cluster_tps,
                    Label::Estimated,
                    Some(&tps_source),
                ));
            }
        }
    }

    let weight_source = format!(
        "{} bytes / {num_gpus} TP ranks",
        fmt_u64(total_weight_bytes)
    );
    let per_gpu_source = format!(
        "{memory_bandwidth_gbps} GB/s × {:.0}% util / {} weight bytes",
        bw_utilization * 100.0,
        fmt_u64(weight_per_gpu)
    );
    let cluster_source = format!(
        "per-GPU × {num_gpus} GPUs × {:.0}% comm × {nvlink_eff:.3} NVLink penalty (NVLink={} GB/s)",
        cluster_comm_efficiency * 100.0,
        gpu.nvlink_bandwidth_gbps
    );

    DecodeEstimate {
        active_weight_bytes_per_gpu: AnnotatedValue::new(
            weight_per_gpu,
            Label::Estimated,
            Some(&weight_source),
        ),
        per_gpu_tokens_per_sec: AnnotatedValue::new(
            per_gpu_tps,
            Label::Estimated,
            Some(&per_gpu_source),
        ),
        cluster_tokens_per_sec: AnnotatedValue::new(
            cluster_tps,
            Label::Estimated,
            Some(&cluster_source),
        ),
        bw_utilization,
        cluster_comm_efficiency,
        moe_active_weight_bytes_per_gpu: moe_active_weight,
        moe_active_tokens_per_sec: moe_active_tps,
    }
}

fn unknown_decode(bw_utilization: f64, cluster_comm_efficiency: f64) -> DecodeEstimate {
    DecodeEstimate {
        active_weight_bytes_per_gpu: AnnotatedValue::new(
            0,
            Label::Unknown,
            Some("GPU memory_bandwidth_gbps not in database"),
        ),
        per_gpu_tokens_per_sec: AnnotatedValue::new(
            0.0,
            Label::Unknown,
            Some("GPU memory_bandwidth_gbps not in database"),
        ),
        cluster_tokens_per_sec: AnnotatedValue::new(
            0.0,
            Label::Unknown,
            Some("GPU memory_bandwidth_gbps not in database"),
        ),
        bw_utilization,
        cluster_comm_efficiency,
        moe_active_weight_bytes_per_gpu: None,
        moe_active_tokens_per_sec: None,
    }
}

fn fmt_u64(value: u64) -> String {
    let text = value.to_string();
    let mut out = String::with_capacity(text.len() + text.len() / 3);
    for (idx, ch) in text.chars().rev().enumerate() {
        if idx > 0 && idx % 3 == 0 {
            out.push(',');
        }
        out.push(ch);
    }
    out.chars().rev().collect()
}

fn fmt_scientific(value: u64) -> String {
    let text = format!("{value:.2e}");
    if let Some((mantissa, exponent)) = text.split_once('e') {
        if !exponent.starts_with(['+', '-']) {
            return format!("{mantissa}e+{exponent}");
        }
    }
    text
}
